### ISSUE TITLE
Infinite buffering on newer ExoPlayer versions

### ISSUE THREAD

**Issue by samuuh** [google/ExoPlayer#8593]
Hi,

We have an issue with more recent versions of exoplayer where stream starts buffering indefinitely on certain platforms. Stream does start properly when video track is changed manually though. 

Issue is seen in Exoplayer versions 2.12.3 and 2.13.1. Same stream works with ExoPlayer version 2.9.3.

The issue happens at least on following emulator configurations in macOS:
Android 9.0 (Google APIs) x86
Android 10.0 (Google APIs) x86
At the moment we have not seen this on real devices, but we have very limited number of test devices so we would like to know the root cause of this before using this version publicly. As mentioned before, the stream has worked with previous ExoPlayer-version, even with these emulators.

This issue is reproducible on demo app. I'll email the media-url and adb bugreport outputs to dev.exoplayer@gmail.com.

**Comment by christosts** [comment]
I can verify it reproduces on Android emulator (tried a Pixel 3 with Android 9). Playback did not start on ExoPlayer r2.13.0 and r2.10.4 but it did play on r2.9.3.

So far, I see the player is frozen because MediaCodec does not return decoded samples, that is although the player is queueing samples for decoding, `MediaCodec.dequeueOutputBuffer()` keeps returning -1 (INFO_TRY_AGAIN_LATER).

One difference I noticed between 2.13.0 and 2.9.3 is that, at least in my tests, 2.13.0 starts playback from the highest resolution/bitrate of the test video which makes the MediaCodec get stuck, whereas 2.9.3 starts from the 640x480 resolution. Interestingly, using the demo app on 2.9.3, when I manually chose the highest resolution, the emulator stopped decoding samples too. I am suspecting the stream appears to work on 2.9.3 because the player does not start from a high resolution, however MediaCodec cannot decode the higher resolution video regardless (it just so happens r2.9.3 does not trigger this immediately). Can you verify this? If you use the demo app on 2.9.3, start playback and then pick the highest video resolutions, does your emulator keep playing or is playback frozen?

I will check internally if there is a known issue with the emulator decoder and the specific stream.

Tests with real devices worked fine.

**Comment by samuuh** [comment]
Thanks a ton for the quick reply!

I can confirm that video works on the 2.9.3 on the first playback; playback is started on lower bitrate video track and then changed to highest. However if playback is started again, the best quality is preselected (apparently bandwith or something is remembered from previous play) and playback will not start. Also I noticed that if I set any stream quality before starting the playback, it will not work even on the first time. With the 2.13.1 I am able to stream the highest quality if I first start playback with Automatic choice, then select lower quality track and switch back to highest. So it looks like the emulator is capable of playing also the higher quality tracks.

Also noticed that on the 2.13.1 this buffering issue can also be reproduced during playback by selecting "none" as video track and after that any other track, the first selection will not work but the next ones will.

Can you see something in the stream that could cause this? We see this only on some of our content. With our streaming team we tried to compare the working and these streams but were unable to figure out what was the relevant difference.

**Comment by christosts** [comment]
Some extra investigation: I think this is an issue when the software H264 decoder decodes the specific content, but I can't understand whether it's an incompatibility of a bug in the decoder. This may affect real devices too: I tested on a device and forced the player to pick up the software decoder and see the same behavior.

I believe it is unlikely that the problem will appear on a device because devices have hardware decoders for H264 and by default ExoPlayer will pick a hardware decoder. However, I will forward this to the platform team to take a look. 

Can  you keep the test content in place while I investigate this further?

**Comment by kpandroid** [comment]
@christosts please check this place - I found that that sorting changes the codec (if two codecs are equal) order and sometimes the wrong codec is used

https://github.com/google/ExoPlayer/blob/4b1e0fa9fc643f96ab5bbf4cec9aa7a79e4bc789/library/core/src/main/java/com/google/android/exoplayer2/video/MediaCodecVideoRenderer.java#L378

**Comment by samuuh** [comment]
> Can you keep the test content in place while I investigate this further?

Content should be available for a good while.

**Comment by christosts** [comment]
For future reference, we have an feature in the roadmap to at least detect cases where the decoder is stuck and raise an error to unblock the app [internal ref b/168800589]

**Comment by christosts** [comment]
@kpandroid You placed a similar issue on #8542 but I don't think the issue reported on either issues is traced to codec re-ordering. 

In the issue reported here, I believe there is a bug in the software decoder which is triggered on the emulator and also on a device, if the device picks the software decoder (which is not the standard behavior of ExoPlayer).

As I wrote on #8542, if you're seeing the software decoder being picked over the hardware decoder on a specific device, please file a separate issue.

**Comment by christosts** [comment]
for future reference [internal ref b/180725591]

**Comment by samuuh** [comment]
Hi, I noticed that tags are now updated as "bug: in platform". Do you have any more information on this? Is there any way we can work around this? The content is playing if some conditions have been met. It is unclear to me what these conditions might be, all the tracks on that provided sample do seem work on the emulator after manual track change.

**Comment by christosts** [comment]
The label is for internal tracking. I put this label because, at the moment, the issue seems to be related to Android's software decoder and not ExoPlayer. But I do not have more on this yet.

**Comment by christosts** [comment]
I got an update. The decoder team replied that

```
the clip has two sets of SPS where num_reorder_frames field in VUI is different (one of the SPS says 1 , other one says 0).
In c2 plugins for s/w decoders this value is used to minimize the output delay and this is expected to be consistent in all the SPS sent to the decoder. Whenever there is a change, decoder is reset and further decoding starts afresh.
In this case since two SPS are sent together, but with differing values for num_reorder_frames, s/w decoder is unable to handle this.

If the content creator can take a look and find out why num_reorder_frames is different between SPS that are sent together and if it is possible to fix that, then that would solve the problem.
```

If you have control of the encoding process, maybe try re-encoding the content with different params? I'm afraid I won't be able to assist more as is outside my expertise. Also, this is not an ExoPlayer specific issue, so I'll close this issue.

**Comment by huseyinunified** [comment]
Hi @christosts

We have some inconsistent outputs from our tests. Could you please answer the questions below which will be very useful to continue investigating?

- Multiple sample entries are allowed so why is this a problem for Exoplayer? Also, why is it only occur with older devices? (We both tested that Pixel 3A works just fine.)

- We don't see any error before the app goes back to the main menu after the stream starts. Why it is not throw an error if something is wrong rather than just staying on buffering?

- We did not see in the logs that the Exoplayer chokes on the multiple sample set, could you please show the relevance in the logs? 

Thanks!

**Comment by christosts** [comment]
> * Multiple sample entries are allowed so why is this a problem for Exoplayer? Also, why is it only occur with older devices? (We both tested that Pixel 3A works just fine.)

The problem is not with  ExoPlayer but with the software decoder which could not decode the specific content provided to us. This situation should not happen on a device because _in most cases_ the player will pick up the device's hardware decoder and not the software decoder.

> * We don't see any error before the app goes back to the main menu after the stream starts. Why it is not throw an error if something is wrong rather than just staying on buffering?

There is no error because the decoder or the platform do not raise any error to the player. The player does not have a mechanism to 'detect' if the platform is stuck. After seeing this behavior, we will schedule to implement a feature that detects if the platform is stuck (i.e. not producing decoded frames and not reporting an error either). To some extent, the player would have to make a guess though and we will need to tune such a feature to avoid any false positives. Therefore, I don't expect such an error-detecting mechanism to arrive very soon. [internal ref b/168800589]

> * We did not see in the logs that the Exoplayer chokes on the multiple sample set, could you please show the relevance in the logs?

To my knowledge, there are no related logs to point to this. The root cause was identified after playing and the stream that was provided with us and debugging the decoder.
